javascript/ql/src/semmle/javascript/dependencies/SemVer.qll

```diff
@@ -0,0 +1,103 @@
+
+/**
+ * Provides classes for working SemVer (Semantic Versioning).
+ */
+
+import semmle.javascript.dependencies.Dependencies
+
+/**
+ * A SemVer-formatted version string in a dependency.
+ */
+class DependencySemVer extends string {
+  Dependency dep;
+
+  int major;
+
+  int minor;
+
+  int patch;
+
+  string ext;
+
+  DependencySemVer() {
+    dep.info(_, this) and
+    exists(string pattern |
+      pattern = "(\\d+)(?:\\.(\\d+)(?:\\.(\\d+))?)?(.*)" and
+      regexpMatch(pattern) and
+      major = this.regexpCapture(pattern, 1).toInt() and
+      (
+        if exists(this.regexpCapture(pattern, 2))
```

Comment:
Could you simplify the logic here by instead using a regex like `(\d+)((?:\.\d+)?)((?:\.\d+)?)(.*)`, i.e., all capture groups are always defined, plus a helper predicate that converts `".nnn"` to `nnn` and `""` to `0`? (Or indeed use `.splitAt(".", ...)` instead of the regex.)

Reply:
Of course, using @asger-semmle's suggestion would be even better.